iroha/src/sumeragi.rs

```diff
@@ -656,343 +659,561 @@ impl<Q: QueueTrait, G: GenesisNetworkTrait, W: WorldTrait> Sumeragi<Q, G, W> {
         self.transactions_awaiting_created_block.clear();
         self.transactions_awaiting_receipts.clear();
-        let previous_role = self.network_topology.role(&self.peer_id);
-        self.number_of_view_changes += 1;
-        if self.number_of_view_changes <= self.n_topology_shifts_before_reshuffle {
-            self.network_topology.shift_peers_by_one();
-        } else {
-            self.network_topology.sort_peers_by_hash_and_counter(
-                Some(self.latest_block_hash),
-                self.number_of_view_changes,
-            )
-        }
+        let previous_role = self.topology.role(&self.peer_id);
+        self.topology.apply_view_change();
         *self.voting_block.write().await = None;
         iroha_logger::info!(
             "{} - {:?} - Changing view at block with hash {}. New role: {:?}. Number of view changes (including this): {}",
             self.peer_id.address,
             previous_role,
-            self.latest_block_hash,
-            self.network_topology.role(&self.peer_id),
-            self.number_of_view_changes,
+            self.latest_block_hash(),
+            self.topology.role(&self.peer_id),
+            self.number_of_view_changes(),
         );
     }
 
     /// If this peer is a leader in this round.
     pub fn is_leader(&self) -> bool {
-        self.network_topology.role(&self.peer_id) == Role::Leader
+        self.topology.role(&self.peer_id) == Role::Leader
     }
 
     /// Returns current network topology or genesis specific one, if the `block` is a genesis block.
-    pub fn network_topology_current_or_genesis(
-        &self,
-        header: &BlockHeader,
-    ) -> InitializedNetworkTopology {
+    pub fn network_topology_current_or_genesis(&self, header: &BlockHeader) -> Topology {
         if header.is_genesis() && self.block_height == 0 {
             if let Some(genesis_topology) = &header.genesis_topology {
                 iroha_logger::info!("Using network topology from genesis block.");
                 return genesis_topology.clone();
             }
         }
 
-        self.network_topology.clone()
+        self.topology.clone()
     }
 }
 
 impl<Q: QueueTrait, G: GenesisNetworkTrait, W: WorldTrait> Debug for Sumeragi<Q, G, W> {
     fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
         f.debug_struct("Sumeragi")
             .field("public_key", &self.key_pair.public_key)
-            .field("network_topology", &self.network_topology)
+            .field("network_topology", &self.topology)
             .field("peer_id", &self.peer_id)
             .field("voting_block", &self.voting_block)
             .finish()
     }
 }
 
-/// Uninitialized `NetworkTopology`, use only for construction.
-/// Call `init` to get `InitializedNetworkTopology` and access all other methods.
-#[derive(Debug)]
-pub struct NetworkTopology {
-    peers: HashSet<PeerId>,
-    max_faults: u32,
-    block_hash: Option<Hash>,
-}
+pub mod network_topology {
```

Comment:
Btw, could you move it to separate file? It will make nice hirarchy and it will be way easier to navigate in project if we move all nested mods to their own files